blog/2020-10-01-hs-to-coq/2020-10-01-hs-to-coq.md

```diff
@@ -0,0 +1,1084 @@
+to get the underlying assumption that `xs = front ++ rev back`, for some
+`front` and `back` such that `q = MkQueue front back`.
+This decomposition is performed by the "intro pattern" `[front back Eqn_xs]`,
+giving us an assumption `Eqn_xs : (xs = front ++ rev back)`
+(i.e., an equation named `Eqn_xs`).
+`q` gets substituted with `MkQueue front back`, so we can forget it,
+with the intro pattern `_`, since it is no longer needed.
+
+After substituting `q`, the goal becomes:
+
+```coq
+(* Current goal *)
+models (xs ++ [x]) (push x (MkQueue front back))
+```
+
+####### 2. `cbn`
+
+Since `push` is defined by pattern-matching on its second argument, we can simplify
+that expression, which is what the tactic `cbn` does. That simplifies to:
+
+```coq
+(* Current goal *)
+models (xs ++ [x]) (MkQueue front (x :: back))
+```
+
+####### 3. `apply models_MkQueue`
+
+Now that the second argument of `models` is of the form `MkQueue FRONT BACK`,
+we can again use the definition of `models`, embodied by the theorem
+`models_MkQueue`, to transform the goal into an equation.
+
+```coq
+(* Current goal *)
+(xs ++ [x]) = (front ++ rev (x :: back))
+```
+
+####### 4. `cbn`, again
+
+The function `rev` is also defined by pattern-matching, so the natural next step
+is to simplify again, using `cbn`. Note that `rev` uses the naive definition to
+reverse a list, which is slow at run time but convenient for proofs.
+
+```coq
+(* Current goal *)
+(xs ++ [x]) = (front ++ (rev back ++ [x]))
+```
+
+####### 5. `rewrite Eqn_xs`
+
+We have an assumption `xs = (front ++ rev back)`, named `Eqn_xs`.
+The `rewrite` tactic uses such an equality to substitute one side with
+the other.
+
+```coq
+(* Current goal *)
+((front ++ rev back) ++ [x]) = (front ++ (rev back ++ [x]))
+```
+
+####### 6. `rewrite app_assoc`
+
+Finally, by associativity, we can rewrite one side to become identical to the
+other side.
+
+```coq
+(* Current goal *)
+((front ++ rev back) ++ [x]) = ((front ++ rev back) ++ [x])
+```
+
+####### 7. `reflexivity`
+
+Both sides are identical. The goal is solved by reflexivity.
+
+That concludes the proof of `models_push`. `Qed.`
+
+##### Verify `pop`
+
+The specification and proof for `pop` is trickier because its result
+is an optional value. The following theorem says that if `q` is
+modelled by a nonempty list `y :: xs`, then `pop q` must succeed
+with `Some` result. The first component of the result is `y`.
+The second component is a queue `q'`, which must be modelled by the rest of the
+list `xs`.
+The list `q'` is quantified existentially: it is an output determined by the
+implementation.
+In contrast, universally quantified parameters are those chosen by clients of the
+implementation.
+
+<div class="tuto">
+```alectryon
+Theorem models_pop {a} : forall (xs : list a) (y : a) (q : Queue a),
+    models (y :: xs) q ->
+    exists q',
+      pop q = Some (y, q') /\
+      models xs q'.
+```
```

Comment:
This only specifies the non-empty case, right?

Reply:
Right. I added a mention of that, and made addressing it an exercise for the reader.

Reply:
Always good to have exercises 😀